### ISSUE TITLE
[initial review] Pseudo-random number generator (Mersenne twister), plus Nat32

### ISSUE THREAD

**Issue by atacratic** [unisonweb/base_v1#31]
This PR gives people access to pseudo-random numbers.  Plus it includes a couple of other things I needed along the way, to achieve that.  I'm after an initial review now, even though I still have a bit more work to do anyway before it can be merged, which I will get on with in parallel.  

[edited 28-Apr; still not complete; would still appreciate an initial review]

## What's included:
- a bunch of stuff under random.mersenne
- Nat32 and a bunch of associated operators, for 32-bit arithmetic
- a few bits and pieces:
  - boolean xor
  - 'iterate', similar to the existing 'repeat'.
  - `until : (a ->{e} Boolean) -> '{e} a ->{e} a` 
  - `all, any, none :  (a ->{𝕖} Boolean) ->{𝕖} [a] ->{𝕖} Boolean`
  - `Nat.max`, `Nat.min`, `Nat.inRange`; and similar for `Int`

## Rationale for the Mersenne twister PRNG
The temptingly-simply one-liner linear congruential generators that sometimes get used for generating pseudo-random numbers are not sound for statistical use.  

The Mersenne twister algorithm *is* sound for this use, and indeed is the most common choice in software libraries, see https://en.wikipedia.org/wiki/Mersenne_Twister#Adoption_in_software_systems

It's not sound for cryptographic use, as is prominently noted in the docs.  

Since this is a pseudo-random number generator, it is suitable for use in pure programs.  Exposing other sources of true randomness through `IO` (with the help of the OS and/or RDRAND instruction) is out of scope for now.    

## Things to highlight to reviewers
- I've wrapped `Nat` in a `Nat32` type, so I can keep the top 32 bits zeroed out as I do math on the bottom 32 bits, without having to think too hard.  The `Nat32` stuff would all look very different in a language with newtypes and typeclasses.  I've taken some inspiration from haskell's [`Data.bits`](https://hackage.haskell.org/package/base-4.12.0.0/docs/Data-Bits.html) anyway, just in terms of naming.
- My definition `type Nat32 = Nat32 Nat` does not specify `unique` at the moment.  Should it?
- Is the name I've used for `mkNat32 : Nat -> Optional Nat32` OK?  Is there a convention here?
- The name I've chosen for `iterate` (note Haskell has a different function with the same name), and note also the similarity to `repeat`.
- Is trunk.random.mersenne.doc named OK?
- `Nat.inRange` and `Int.inRange` are inclusive on the left but exclusive on the right.
- The API uses the existing `Ask` ability.  Maybe it would be better to have a (`unique`) ability called `Random`, with the same signature as `Ask`, to express intent.  Or `Choice` maybe.  (Note the existing `Ask.provide` function uses a handler that returns the same value on every call to `ask`, so the semantics are quite different to what the PRNG is doing.)

## Loose end / discussion point
- If I read it correctly (which I might well not have), I think the test library in base does its property-based testing by enumerating possible values from small to large (at least for the Nat domain, say).  Would it be good if it used a PRNG to yield more far-flung values?

## Things I still need to do

```
.me.base> display trunk.random.mersenne.TODO
  
  TODO...
   - debug failing test
   - add a test against a list of expected outputs
   - add Ask handler for Floats
   - comment that next assumes the state is seeded, or will crash
   - comment re considerations for seeding (time, location), helper to seed by time,
  default seed (replace that in mersenne.doc.examples)

```
- move `until` to the right place
- delete patches
- update screenshot below

## User docs

The main doc looks like this... [TODO update pic]

![image](https://user-images.githubusercontent.com/21991324/80320040-d226d700-880b-11ea-98de-7900bb752239.png)

## Code review

```
  The changes summarized below are available for you to review, using the following command:
  
    pull-request.load https://github.com/unisonweb/base https://github.com/atacratic/unisonweb-base.git
  
  Updates:
  
     patch trunk.patch (added 5 updates)
  
  Added definitions:
  
     type trunk.Nat32 (+3 metadata)
     type trunk.random.mersenne.State (+3 metadata)
     trunk.Nat32.Nat32                                  : Nat -> Nat32 (+3 metadata)
     trunk.random.mersenne.State.State                  : [Nat32] -> State (+2 metadata)
     trunk.Int.inRange                                  : Int
                                                        -> Int
                                                        -> Int
                                                        -> Boolean (+3 metadata)
     trunk.Int.inRange.doc                              : Doc (+2 metadata)
     trunk.Int.max                                      : Int -> Int -> Int (+2 metadata)
     trunk.Int.min                                      : Int -> Int -> Int (+2 metadata)
     trunk.Nat.inRange                                  : Nat
                                                        -> Nat
                                                        -> Nat
                                                        -> Boolean (+3 metadata)
     trunk.Nat.inRange.doc                              : Doc (+2 metadata)
     trunk.Nat.max                                      : Nat -> Nat -> Nat (+2 metadata)
     trunk.Nat.min                                      : Nat -> Nat -> Nat (+2 metadata)
     trunk.Nat32.*                                      : Nat32 -> Nat32 -> Nat32 (+2 metadata)
     trunk.Nat32.+                                      : Nat32 -> Nat32 -> Nat32 (+2 metadata)
     trunk.Nat32.<                                      : Nat32 -> Nat32 -> Boolean (+2 metadata)
     trunk.Nat32.<=                                     : Nat32 -> Nat32 -> Boolean (+2 metadata)
     trunk.Nat32.>                                      : Nat32 -> Nat32 -> Boolean (+2 metadata)
     trunk.Nat32.>=                                     : Nat32 -> Nat32 -> Boolean (+2 metadata)
     trunk.Nat32.Nat32.doc                              : Doc (+2 metadata)
     trunk.Nat32.and                                    : Nat32 -> Nat32 -> Nat32 (+3 metadata)
     trunk.Nat32.and.doc                                : Doc (+2 metadata)
     trunk.Nat32.bit                                    : Nat -> Nat32 (+3 metadata)
     trunk.Nat32.bit.doc                                : Doc (+2 metadata)
     trunk.Nat32.bit.test1                              : [Result] (+3 metadata)
     trunk.Nat32.bit.test2                              : [Result] (+3 metadata)
     trunk.Nat32.bit.test3                              : [Result] (+3 metadata)
     trunk.Nat32.clearBit                               : Nat32 -> Nat -> Nat32 (+4 metadata)
     trunk.Nat32.clearBit.doc                           : Doc (+3 metadata)
     trunk.Nat32.clearBit.test1                         : [Result] (+4 metadata)
     trunk.Nat32.clearBit.test2                         : [Result] (+4 metadata)
     trunk.Nat32.complement                             : Nat32 -> Nat32 (+3 metadata)
     trunk.Nat32.complement.doc                         : Doc (+2 metadata)
     trunk.Nat32.doc                                    : Doc (+3 metadata)
     trunk.Nat32.isEven                                 : Nat32 -> Boolean (+2 metadata)
     trunk.Nat32.maxNat32                               : Nat32 (+2 metadata)
     trunk.Nat32.maxNat32.test                          : [Result] (+3 metadata)
     trunk.Nat32.mkNat32                                : Nat -> Optional Nat32 (+2 metadata)
     trunk.Nat32.multiply.prop1                         : [Result] (+3 metadata)
     trunk.Nat32.plus.prop1                             : [Result] (+3 metadata)
     trunk.Nat32.shiftLeft                              : Nat32 -> Nat -> Nat32 (+2 metadata)
     trunk.Nat32.shiftLeft.prop1                        : [Result] (+3 metadata)
     trunk.Nat32.shiftRight                             : Nat32 -> Nat -> Nat32 (+2 metadata)
     trunk.Nat32.shiftRight.prop1                       : [Result] (+3 metadata)
     trunk.Nat32.toNat                                  : Nat32 -> Nat (+2 metadata)
     trunk.Nat32.truncate32                             : Nat -> Nat32 (+2 metadata)
     trunk.Nat32.truncate32.prop1                       : [Result] (+3 metadata)
     trunk.Nat32.truncate32.prop2                       : [Result] (+3 metadata)
     trunk.Nat32.xor                                    : Nat32 -> Nat32 -> Nat32 (+3 metadata)
     trunk.Nat32.xor.doc                                : Doc (+2 metadata)
     trunk.all                                          : (a ->{𝕖} Boolean)
                                                        ->{𝕖} [a]
                                                        ->{𝕖} Boolean (+2 metadata)
     trunk.any                                          : (a ->{𝕖} Boolean)
                                                        ->{𝕖} [a]
                                                        ->{𝕖} Boolean (+2 metadata)
     trunk.iterate                                      : Nat -> '{e} a ->{e} [a] (+3 metadata)
     trunk.iterate.doc                                  : Doc (+2 metadata)
     trunk.iterate.test                                 : [Result] (+3 metadata)
     trunk.metadata.authors.atacratic                   : Author
     trunk.metadata.authors.atacratic.guid              : GUID
     trunk.metadata.copyrightHolders.atacratic          : CopyrightHolder
     trunk.metadata.licenses.atacratic.mit_2020         : License
     trunk.none                                         : (a ->{𝕖} Boolean)
                                                        ->{𝕖} [a]
                                                        ->{𝕖} Boolean (+2 metadata)
     trunk.random.mersenne.State.doc                    : Doc (+2 metadata)
     trunk.random.mersenne.TODO                         : Doc (+2 metadata)
     trunk.random.mersenne.collect                      : '{Store State} x
                                                        -> Nat32
                                                        -> Nat
                                                        -> [x] (+3 metadata)
     trunk.random.mersenne.collect.doc                  : Doc (+3 metadata)
     trunk.random.mersenne.defaultSeed                  : Nat32 (+3 metadata)
     trunk.random.mersenne.defaultSeed.doc              : Doc (+2 metadata)
     trunk.random.mersenne.doc                          : Doc (+3 metadata)
     trunk.random.mersenne.doc.example1                 : [Nat] (+2 metadata)
     trunk.random.mersenne.doc.example2                 : Nat (+2 metadata)
     trunk.random.mersenne.handler                      : '{Store State} n
                                                        -> Nat32
                                                        -> Request (Ask n) a
                                                        -> a (+3 metadata)
     trunk.random.mersenne.handler.doc                  : Doc (+2 metadata)
     trunk.random.mersenne.int.next                     : '{Store State} Int (+3 metadata)
     trunk.random.mersenne.int.next.doc                 : Doc (+2 metadata)
     trunk.random.mersenne.int.nextFromRange            : Int
                                                        -> Int
                                                        -> '{Store State} Int (+3 metadata)
     trunk.random.mersenne.int.nextFromRange.doc        : Doc (+2 metadata)
     trunk.random.mersenne.int.nextFromRange.prop       : [Result] (+3 metadata)
     trunk.random.mersenne.nat.next                     : '{Store State} Nat (+3 metadata)
     trunk.random.mersenne.nat.next.doc                 : Doc (+2 metadata)
     trunk.random.mersenne.nat.nextFromRange            : Nat
                                                        -> Nat
                                                        -> '{Store State} Nat (+3 metadata)
     trunk.random.mersenne.nat.nextFromRange.doc        : Doc (+2 metadata)
     trunk.random.mersenne.nat.nextFromRange.doc.caveat : Doc (+2 metadata)
     trunk.random.mersenne.nat.nextFromRange.prop       : [Result] (+3 metadata)
     trunk.random.mersenne.nat32                        : Nat32
                                                        -> '{Ask Nat32} a
                                                        -> a (+3 metadata)
     trunk.random.mersenne.nat32.doc                    : Doc (+3 metadata)
     trunk.random.mersenne.nat32.handler                : Nat32
                                                        -> Request (Ask Nat32) a
                                                        -> a (+2 metadata)
     trunk.random.mersenne.next                         : '{Store State} Nat32 (+4 metadata)
     trunk.random.mersenne.next.doc                     : Doc (+3 metadata)
     trunk.random.mersenne.provide                      : '{Store State} n
                                                        -> Nat32
                                                        -> '{Ask n} a
                                                        -> a (+3 metadata)
     trunk.random.mersenne.provide.doc                  : Doc (+3 metadata)
     trunk.random.mersenne.seed                         : Nat32 ->{Store State} () (+3 metadata)
     trunk.random.mersenne.seed.doc                     : Doc (+3 metadata)
     trunk.random.mersenne.seed.tests.size              : [Result] (+4 metadata)
     trunk.random.mersenne.test.comparison              : ∀ (). () ->{IO} () (+4 metadata)
     trunk.random.mersenne.test.comparison.doc          : Doc (+3 metadata)
     trunk.random.mersenne.test.printOut                : [Nat] -> Nat ->{IO} () (+3 metadata)
     trunk.random.mersenne.test.printOut.doc            : Doc (+3 metadata)
     trunk.xor                                          : Boolean
                                                        -> Boolean
                                                        -> Boolean (+3 metadata)
     trunk.xor.doc                                      : Doc (+2 metadata)
     trunk.xor.test                                     : [Result] (+3 metadata)
     until                                              : (a ->{e} Boolean)
                                                        -> '{e} a
                                                        ->{e} a (+3 metadata)
     until.doc                                          : Doc (+2 metadata)
  
     patch trunk.random.patch (added 11 updates)

```

**Comment by aryairani** [comment]
@atacratic I haven't dug into this PR at all, but I'm excited about it.  Thanks for putting it together!

**Comment by aryairani** [comment]
> * My definition `type Nat32 = Nat32 Nat` does not specify `unique` at the moment.  Should it?

If you want these to be different, then yes:
```
type Nat32 = Nat32 Nat
type Nat16 = Nat16 Nat
type Foo = Bar Nat
```
Basically anything that's not defined _by_ its type signature (which in this case is just `Nat`) should be `unique`.

**Comment by aryairani** [comment]
I haven't opened it in ucm, but here's a superficial review based on what I can see from Github — thanks for including questions for review.

> patch trunk.patch (added 5 updates)
> patch trunk.random.patch (added 11 updates)

I haven't looked at their contents, I'm guessing these changes were included in the PR accidentally?  Apologies that the tooling around patches isn't where we want it yet, necessitating such manual cleanup.  As a workaround, you could develop `Nat32` or whatever feature outside of `base`, and then `alias.many` everything into place right before creating the PR.

> Is the name I've used for mkNat32 : Nat -> Optional Nat32 OK? Is there a convention here?

I'd suggest `Nat32.fromNat`

> Is trunk.random.mersenne.doc named OK?

Seems okay to me — @pchiusano ?

> Nat.inRange and Int.inRange are inclusive on the left but exclusive on the right.

Yeah this is always tough, but I think this is good for computery reasons.  Just need docs and consistency.

> The API uses the existing Ask ability. Maybe it would be better to have a (unique) ability called Random, with the same signature as Ask, to express intent. Or Choice maybe. (Note the existing Ask.provide function uses a handler that returns the same value on every call to ask, so the semantics are quite different to what the PRNG is doing.)

Yeah, `unique` types and abilities are about semantics, and semantics seem super important here.

**Comment by atacratic** [comment]
Thanks @aryairani 😄 

I'm not totally clear what you're thinking on `unique`.  Is it 100% definite for you that we want `type Nat32 = Nat32 Nat` to be unique?  It's not totally obvious to me that it would be bad for (say) `Nat32.<` and `Nat16.<` to hash the same?  

Also are you saying that we should definitely have a `unique` version of `Ask` (called `Random` maybe)?  And in that case is it right that `Ask` is not unique?  (I'm assuming `Ask` is not declared unique already, I didn't actually check)

**Comment by pchiusano** [comment]
Just took very quick look based on the API. I think I’d also use a unique ability rather than just Ask. I think the idea of using something Ask-like is nice as it means the handler is free to use whatever state it wants for the RNG, which is nice and general.

I’d also like to see a splittable RNG, which is more useful for functional programs IMO.. and might be interesting to think about how to work that into the API. Though not bad to have both I suppose. Perhaps this as inspiration, which as a bonus works directly with 64 bit numbers: https://hackage.haskell.org/package/splitmix

Minor thing: PR and the diff is a little cleaner against trunk of base rather than the root.

**Comment by atacratic** [comment]
Hmm well now I've discovered https://www.pcg-random.org/ and am wishing I'd found it sooner...  seems like pcg64 is even simpler in code terms than Mersenne Twister, has a bunch of improved properties.  Gets a decent rating on this page too: https://rust-random.github.io/book/guide-rngs.html 

I haven't worked out whether it is splittable in the sense that SplitMix is, and how that relates to 'jump-ahead' and 'streams', but will keep looking.  https://www.pcg-random.org does have some criticisms of SplitMix so not sure that one's a shoo-in.  

(Also some decent info at https://www.codeproject.com/Articles/1083372/Random-Number-Generator-Recommendations-for-Applic?msg=5519945#xx5519945xx )

I think for now I'll put this mersenne-twister implementation on github as a standalone lib, do a new base PR for the misc helpers it included, and come back at some point hopefully fairly soon to implementing maybe pcg64 for base.

I think fundamentally having a fairly statistically sound PRNG in base is desirable, as long as its simple to implement.  But within those constraints it's worth going with an algorithm that won't make people who know about these things roll their eyes.

'Days of coding can save you hours of thinking'... or minutes of online research  🤷 

Thanks @pchiusano and @aryairani for the comments, I'll factor those in next time round.

**Comment by pchiusano** [comment]
@atacratic cool, also be sure to add your library here https://www.unisonweb.org/docs/libraries/ and it'll show up in the site search results. As of this writing, you'll be the first besides `base`! I encourage this even if you're still just tinkering around - you can just tag it appropriately so people know it's still in flux.

<img width="674" alt="Screen Shot 2020-04-23 at 6 37 58 PM" src="https://user-images.githubusercontent.com/11074/80550845-67060d80-898f-11ea-98ca-f139685a77d7.png">
